### PR TITLE
Implement VK_EXT_conditional_rendering

### LINKED ISSUE #1180: Support for VK_EXT_conditional_rendering
Hello,

I'm currently having some trouble implementing D3D11 Predication in DXVK using the `VK_EXT_conditional_rendering` extension, so I'd like to ask for Renderdoc to support this extension if possible.

A demo that could be used for testing purposes is the "Conditional Rendering" sample in Sascha Willems' Vulkan samples.

### PR DESCRIPTION
**Pull request by werman**
Implementation is similar to D3D11:
- Conditional rendering is skipped during replay
- Predicate value and result are displayed in the raster state

That's my attempt to implement VK_EXT_conditional_rendering, it's not in its final state since I need to add several serialization methods for VKPipe::ConditionalRendering, and test it some more.

It was tested on Sascha Willems' Vulkan sample on Linux with Amd and Intel GPUs.

Fixes #1180

### CODE REVIEW

**Comment by werman** [comment]
I believe now it's ready for review. 
Also I found that in D3D11 predicate state is not exported to HTML so I decided to do the same for Vulkan.

File: renderdoc/api/replay/vk_pipestate.h [hunk 1]
```diff
@@ -887,6 +887,30 @@ struct ImageData
...
+  ResourceId bufferId;
+
+  DOCUMENT("The byte offset into buffer where the predicate is located.");
+  uint64_t byteOffset = 0;
+
+  DOCUMENT("Specifies the behavior of conditional rendering.");
+  uint32_t flags = 0;
+
+  DOCUMENT("Value of the predicate.");
+  uint32_t value = 0;
```
Comment (baldurk) [inline renderdoc/api/replay/vk_pipestate.h]: With `flags` I'm not sure we should expose that right now unless there are many other flags, at which point we can add a proper bitfield enum so that the UI can know how to interpret it without needing to look up vulkan headers. We should have something like a `bool isInverted` though to know if we're testing against zero or non-zero.

The `value` from buffer readback should not be here in the pipeline state, buffer contents can be viewed directly in the UI - this struct should just be the bound state.

In the D3D11 pipeline state the `value` is actually more like a reference value and is the equivalent of the `INVERTED` bit. When binding predication you specify whether you're testing against TRUE or FALSE and the predicate itself has a TRUE or FALSE value. So in the D3D11 state we pass back the test value, and then whether it passes or not.

File: qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.ui [hunk 2]
```diff
@@ -1806,6 +1806,67 @@
...
+          <item>
+           <widget class="RDLabel" name="predicate">
+            <property name="frameShape">
+             <enum>QFrame::Box</enum>
+            </property>
+            <property name="margin">
+             <number>4</number>
+            </property>
+           </widget>
+          </item>
```
Comment (baldurk) [inline qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.ui:1821]: We need a field to display the byte offset here. I think we should also have a `View` button similar to shaders, which will open up the buffer viewer with the format `uint`, with size of 4 bytes, at the offset given. Otherwise it's a bit clumsy to find & view the right value in the buffer.

File: renderdoc/driver/vulkan/vk_common.h [hunk 3]
```diff
@@ -936,6 +943,7 @@ DECLARE_DESERIALISE_TYPE(VkTextureLODGatherFormatPropertiesAMD);
 DECLARE_DESERIALISE_TYPE(VkValidationCacheCreateInfoEXT);
 DECLARE_DESERIALISE_TYPE(VkValidationFlagsEXT);
 DECLARE_DESERIALISE_TYPE(VkWriteDescriptorSet);
+DECLARE_DESERIALISE_TYPE(VkConditionalRenderingBeginInfoEXT);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_common.h]: I should definitely document this better, but all of the types need a `DECLARE_DESERIALISE_TYPE` technically. Anything that has a `pNext` chain will need a `Deserialise` implementation in case it's the root of a chain somewhere, and any `Deserialise` implementation needs to be declared in the header, otherwise the default template will be used.

I know that might seem arbitrary since most of these structs aren't used on their own and probably never will be, but it's easier to enforce consistent rules and just handle everything the same, otherwise you can get lost in exceptions and missing cases.

tl;dr all structs with `pNext` chains need `DECLARE_DESERIALISE_TYPE`

File: renderdoc/driver/vulkan/vk_next_chains.cpp [hunk 4]
```diff
@@ -308,6 +310,8 @@ static void AppendModifiedChainedStruct(byte *&tempMem, VkStruct *outputStruct,
                 UnwrapInPlace(out->commandPool));                                                    \
   UNWRAP_STRUCT(VK_STRUCTURE_TYPE_COMMAND_BUFFER_INHERITANCE_INFO, VkCommandBufferInheritanceInfo,   \
                 UnwrapInPlace(out->renderPass), UnwrapInPlace(out->framebuffer));                    \
+  UNWRAP_STRUCT(VK_STRUCTURE_TYPE_COMMAND_BUFFER_INHERITANCE_CONDITIONAL_RENDERING_INFO_EXT,         \
+                VkCommandBufferInheritanceConditionalRenderingInfoEXT);                              \
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_next_chains.cpp]: Again this is a lack of documentation on my part - because this struct has no members that need to be `UnwrapInPlace`'d it can just be a `COPY_STRUCT`.

Basically the common cases are:

* `COPY_STRUCT` for any struct that doesn't have anything special in it, we just need to copy it onto our patched `pNext` chain.
* `UNWRAP_STRUCT` for any struct where it just has a handle member that we can unwrap after copying, with no other special handling.

Then structs that have e.g. arrays of handles where we need to allocate room for the array and unwrap into it, they have implementations for each.

File: renderdoc/driver/vulkan/vk_serialise.cpp [hunk 5]
```diff
@@ -565,6 +565,14 @@ SERIALISE_VK_HANDLES();
   PNEXT_STRUCT(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_SHADER_IMAGE_FOOTPRINT_FEATURES_NV,                  \
                VkPhysicalDeviceShaderImageFootprintFeaturesNV)                                        \
                                                                                                       \
+  /* VK_EXT_conditional_rendering */                                                                  \
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_serialise.cpp]: Super minor nit-pick - I've tried to sort these extensions within the macro so it's easier to find where each one is. So first within the supported structs I have an alphabetical list of extensions, then second within unsupported structs. This one would go further up between `VK_EXT_astc_decode_mode` and `VK_EXT_conservative_rasterization`.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp [hunk 6]
```diff
@@ -4720,6 +4720,109 @@ void WrappedVulkan::vkCmdEndQueryIndexedEXT(VkCommandBuffer commandBuffer, VkQue
...
+    record->MarkResourceFrameReferenced(buf->GetResourceID(), eFrameRef_Read);
+    record->MarkResourceFrameReferenced(buf->baseResource, eFrameRef_Read);
+
+    if(IsActiveCapturing(m_State))
+      GetResourceManager()->MarkResourceFrameReferenced(buf->GetResourceID(), eFrameRef_Read);
+  }
+}
+
+template <typename SerialiserType>
+bool WrappedVulkan::Serialise_vkCmdBeginConditionalRenderingEXT(
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: Another really minor nit-pick, for no good reason at all I've always placed the `Serialise` implementations above the functions themselves. I don't know why, but at this point it's the way things are and it's better to be consistent.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp [hunk 7]
```diff
@@ -4720,6 +4720,109 @@ void WrappedVulkan::vkCmdEndQueryIndexedEXT(VkCommandBuffer commandBuffer, VkQue
...
+
+    record->AddChunk(scope.Get());
+
+    VkResourceRecord *buf = GetRecord(pConditionalRenderingBegin->buffer);
+
+    record->MarkResourceFrameReferenced(buf->GetResourceID(), eFrameRef_Read);
+    record->MarkResourceFrameReferenced(buf->baseResource, eFrameRef_Read);
+
+    if(IsActiveCapturing(m_State))
+      GetResourceManager()->MarkResourceFrameReferenced(buf->GetResourceID(), eFrameRef_Read);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: This call to `MarkResourceFrameReferenced` shouldn't be needed, the ones above in the command buffer's record should be sufficient. Also marking referenced directly in the resource manager side-steps tracking of which command buffers actually get submitted, so it's to be avoided in the general case.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp [hunk 8]
```diff
@@ -4720,6 +4720,109 @@ void WrappedVulkan::vkCmdEndQueryIndexedEXT(VkCommandBuffer commandBuffer, VkQue
...
+
+  SERIALISE_CHECK_READ_ERRORS();
+
+  if(IsReplayingAndReading())
+  {
+    m_LastCmdBufferID = GetResourceManager()->GetOriginalID(GetResID(commandBuffer));
+
+    if(IsActiveReplaying(m_State) && InRerecordRange(m_LastCmdBufferID))
+    {
+      m_RenderState.conditionalRendering.buf = GetResID(BeginInfo.buffer);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: This is a tricky one. Normally for state setting you'd need an extra call to `ShouldUpdateRenderState` to check if the current renderstate should be updated. Otherwise if your state is set in a secondary command buffer, it will be trampled by any later calls in the primary command buffer while we're re-recording and tracking the state. Current renderpass/subpass/framebuffer is special because that can't be set in the secondary buffer so it's set by the primary.

In this case the conditional rendering can either be set independently in the secondary command buffer _or_ inherited, which I don't think happens with anything else (either state is entirely inherited or entirely reset).

I _think_ the solution would be to do this:
1. This `m_RenderState` is protected by `if(ShouldUpdateRenderState(m_LastCmdBufferID))` which will handle the simple cases correctly - everything but inheritance.
2. Track conditional rendering state in the `m_BakedCmdBufferInfo[m_LastCmdBufferID].state` struct which tracks a subset of the full state per-command buffer instead of globally. Similar to how renderpass/framebuffer is currently.
3. When it comes to `vkCmdExecuteCommands`, if one of the command buffers is one we're partially executing **and** is inheriting conditional rendering state, then copy the `m_BakedCmdBufferInfo[m_LastCmdBufferID].state` conditional state over the `m_RenderState` conditional state. The spec says conditional rendering must be started and stopped in the same command buffer, so we don't have to worry about the state changing within a secondary if it's inherited.

I know that's pretty in-depth :smile:. If you want, in this PR you can just do part 1 and I'll do the rest in a follow-on commit.

**Review by baldurk — CHANGES_REQUESTED** [review]
Thanks for this change, it looks great!

The lack of predication in the D3D11 HTML export is actually a bug, so good find! I'd like it if you would add export to the vulkan HTML export though please. It's up to you if you also add it to D3D11's export, if not then I'll do that myself in a follow-up commit.

I've made a few implementation notes, mostly about improving the UX on the pipeline viewer but there's one little spec-concern on the tracking of conditional render state.

One other thing: I think we should be replaying the conditional rendering state on replay so that things disappear. One of the main reasons I decided not to replay it on D3D11 was because I couldn't replay it consistently. Because D3D11 predicates are opaque objects they are not necessarily queryable to get their current state. That means that depending on whether or not the predicate is filled in the captured frame changes whether it can be replayed correctly. To the user this means that sometimes predicates behave correctly in a capture and sometimes they don't, so I decided the best balance was just to not replay them (especially as they're rarely used).

With vulkan on the other hand because the predication simply points into a buffer that we are already capturing perfectly, we should be able to replay perfectly as well. I think the pipeline state view is fine as you want to be able to show users what's happening, but we should replay it correctly.

This will have a couple of knock-on factors: Similar to transform feedback, you'll need to track if conditional rendering is active and end it if we're partially replaying. You'll also need to add code into the overlays to selectively disable conditional rendering when replaying depending on the overlay, so that the 'highlight drawcall' overlay doesn't get hidden if conditional rendering is failing.

Another factor is that D3D11's documentation says that predicates can be hints, though vaguely, presumably to allow for drivers not syncing between very close fill & use of predicates. On Vulkan there's no such language so I think we'd be spec-invalid if we didn't replay it.

**Comment by baldurk** [comment]
Hopefully the review comments all make sense, let me know if you need clarification on anything at all. The change looks good and I'm happy to have the contribution, just unfortunately this extension hits some non-trivial cases that need to be handled :smile:.

**Comment by werman** [comment]
Thanks for such detailed feedback! Most of it makes sense. At the moment I'm on vacation and will handle it next week, hope it's ok.

File: qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.ui (hunk 2 above)
Comment (werman) [reply qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.ui:1821]: Good idea, changed it to:
image
Will this suffice?

File: renderdoc/driver/vulkan/vk_common.h (hunk 3 above)
Comment (werman) [reply renderdoc/driver/vulkan/vk_common.h]: Understood. There is nothing to change here, right?

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp (hunk 7 above)
Comment (werman) [reply renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: You are right, it is unnecessary.

File: renderdoc/api/replay/vk_pipestate.h (hunk 1 above)
Comment (werman) [reply renderdoc/api/replay/vk_pipestate.h]: Indeed there is no need to expose `flags` and will remove value.

File: qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.ui (hunk 2 above)
Comment (baldurk) [reply qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.ui:1821]: Let's be extra explicit and say `Byte Offset` to be clear in case anyone thinks the offset is in dwords. Otherwise I think that's good.

File: renderdoc/driver/vulkan/vk_common.h (hunk 3 above)
Comment (baldurk) [reply renderdoc/driver/vulkan/vk_common.h]: Ah yes, sorry you have all the `DECLARE_DESERIALISE_TYPE` needed. I misread when going through the code because this one is out of order - I've kept the other structs in alphabetical order to keep things easier to find, I thought this was with the second smaller list of structs that don't have a `pNext` chain but _do_ have some array or dynamic data which needs to be deallocated.

So I think just moving the added entries into alphabetical order in both lists (`DECLARE_REFLECTION_STRUCT` and `DECLARE_DESERIALISE_TYPE`) would do the trick. Note the other two are closer, but not quite alphabetical sorted.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp (hunk 8 above)
Comment (werman) [reply renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: Ughhhh, yes, I have missed this. And it reminded me about what I have missed when I wrote Vulkan CTS tests for this extension...

Will try to implement it the right way.

Comment (werman) [reply renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: I implemented inherited conditional rendering the way you suggested - it works!

**Comment by werman** [comment]
> You'll also need to add code into the overlays to selectively disable conditional rendering when replaying depending on the overlay, so that the 'highlight drawcall' overlay doesn't get hidden if conditional rendering is failing.

Is there any other overlay which should always draw? Maybe "Wireframe Mesh"?

**Comment by werman** [comment]
And one more question: dispatch commands are also affected by conditional rendering, for them I find that placing conditional rendering state into rasterizer is confusing, should I place the info on CS tab in this case?

**Comment by baldurk** [comment]
> Is there any other overlay which should always draw? Maybe "Wireframe Mesh"?

This is a bit of a philosophical question - or at least different people might have different opinions. It came up recently as to whether or not a failure in one part of the pipeline should affect an unrelated overlay. I think ideally the drawcall highlight and wireframe mesh should always render because they're not 'test' overlays. The others that show the pass/fail result should show up but the pass/fail should be affected by predication. So that will mean disabling predication for the pass that renders red, then enabling it for the pass that renders green.

Though this might turn out to be bad behaviour it's at least somewhat consistent :smile:.

> And one more question: dispatch commands are also affected by conditional rendering, for them I find that placing conditional rendering state into rasterizer is confusing, should I place the info on CS tab in this case?

Fair point, the same problem exists on the D3D11 display. I think yes when predication is enabled we should have the same little panel show up in both CS and rasterizer stages. Since 99% of users won't have predication enabled and it disappears entirely when not enabled, I think that's fine.

**Comment by werman** [comment]
> I think ideally the drawcall highlight and wireframe mesh should always render because they're not 'test' overlays. 

Done, though only overlay itself is rendered:

image

> Fair point, the same problem exists on the D3D11 display. I think yes when predication is enabled we should have the same little panel show up in both CS and rasterizer stages. Since 99% of users won't have predication enabled and it disappears entirely when not enabled, I think that's fine.

If it's ok I'll do it in next pull request.

Thanks for the help!

**Review by baldurk — APPROVED** [review]

**Comment by baldurk** [comment]
Thanks again for this, that's not a trivial patch for a first contribution :smile:.

**Comment by werman** [comment]
Thanks for the detailed feedback - it helps!
Implementing the extension in Mesa, writing CTS tests and adding support to Renderdoc is quite an experience.
